File: app/src/main/java/io/ipoli/android/quest/receivers/CompleteQuestReceiver.java
package io.ipoli.android.quest.receivers;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;

import com.squareup.otto.Bus;

import javax.inject.Inject;

import io.ipoli.android.Constants;
import io.ipoli.android.app.App;
import io.ipoli.android.app.events.EventSource;
import io.ipoli.android.quest.events.CompleteQuestRequestEvent;
import io.ipoli.android.quest.persistence.QuestPersistenceService;

/**
 * Created by Venelin Valkov <venelin@curiousily.com>
 * on 10/13/16.
 */
public class CompleteQuestReceiver extends BroadcastReceiver {

    public static final String ACTION_COMPLETE_QUEST = "io.ipoli.android.intent.action.COMPLETE_QUEST";

    @Inject
    QuestPersistenceService questPersistenceService;

    @Inject
    Bus eventBus;

    @Override
    public void onReceive(Context context, Intent intent) {
        App.getAppComponent(context).inject(this);
        PendingResult result = goAsync();

        String questId = intent.getStringExtra(Constants.QUEST_ID_EXTRA_KEY);
        questPersistenceService.findById(questId, quest -> {
            eventBus.post(new CompleteQuestRequestEvent(quest, EventSource.NOTIFICATION));
            result.finish();
        });
    }
}

File: app/src/main/java/mypoli/android/quest/receiver/CompleteQuestReceiver.kt
package mypoli.android.quest.receiver

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import mypoli.android.Constants
import mypoli.android.common.di.SimpleModule
import mypoli.android.myPoliApp
import space.traversal.kapsule.Injects
import space.traversal.kapsule.inject
import space.traversal.kapsule.required

/**
 * Created by Venelin Valkov <venelin@ipoli.io>
 * on 01/22/2018.
 */
class CompleteQuestReceiver : BroadcastReceiver(), Injects<SimpleModule> {

    private val completeQuestUseCase by required { completeQuestUseCase }

    override fun onReceive(context: Context, intent: Intent) {
        inject(myPoliApp.simpleModule(context))
        val questId = intent.getStringExtra(Constants.QUEST_ID_EXTRA_KEY)
        completeQuestUseCase.execute(questId)
    }

}